funcx_sdk/tests/unit/test_login_manager.py

```diff
@@ -159,3 +169,57 @@ def test_get_authorizer(mocker, logman):
 
     with pytest.raises(LookupError):
         logman._get_authorizer("some_resource_server")
+
+
+@pytest.mark.parametrize("has_login", (True, False))
+def test_ensure_logged_in(mocker, logman, has_login):
+    logman.has_login = mock.Mock(return_value=has_login)
+    mock_run_login_flow = mocker.patch(
+        f"{MOCK_BASE}.manager.LoginManager.run_login_flow"
+    )
+
+    logman.ensure_logged_in()
+
+    assert has_login is not mock_run_login_flow.called
+
+
+@pytest.mark.parametrize("tokens", ({"refresh_token": "123"}, {}))
+def test_has_login(mocker, logman, tokens):
+    logman._token_storage.get_token_data = mock.Mock(return_value=tokens)
+    logman._validate_token = mock.Mock(return_value=True)
+    env = {CID_KEY: "some_id", CSC_KEY: "some_secret"}
+
+    assert logman.has_login("funcx") is bool(tokens)
+    with mock.patch.dict(os.environ, env):
+        assert logman.has_login("funcx") is True
+
+
+@pytest.mark.parametrize("auth_error", (True, False))
+def test_requires_login_decorator(mocker, logman, auth_error):
+    mocked_run_login_flow = mocker.patch(
+        f"{MOCK_BASE}.manager.LoginManager.run_login_flow"
+    )
+    mocked_get_funcx_web_client = mocker.patch(
+        f"{MOCK_BASE}.manager.LoginManager.get_funcx_web_client"
+    )
+
+    class MockClient:
+        login_manager = logman
+        funcx_service_address = "127.0.0.1"
+
+        @requires_login
+        def mock_method(self):
+            if auth_error:
+                raise globus_sdk.AuthAPIError(
+                    _fake_http_response(status=400, method="POST")
+                )
+
+    mock_client = MockClient()
+    if auth_error:
+        with pytest.raises(globus_sdk.AuthAPIError):
+            mock_client.mock_method()
+    else:
+        mock_client.mock_method()
+
+    assert mocked_run_login_flow.called is auth_error
+    assert mocked_get_funcx_web_client.called is auth_error
```

Comment:
In the `auth_error` branch, the last thing to verify would be the raison d'être of this PR: that the function gets called a second time after the `run_log_flow` is called.  What would you say to something like this to that end?
```python
    ...

    expected = randomstring()
    mock_method = mock.Mock()
    mock_method.side_effect = [
        expected,
        globus_sdk.AuthAPIError(_fake_http_response(status=400, method="POST")),
        expected
    ]
    mock_method.__name__ = "some_function_name"
    class MockClient:
        login_manager = logman
        funcx_service_address = "127.0.0.1"

        upstream_call = requires_login(mock_method)
    mc = MockClient()

    res = mc.upstream_call(None)  # case: no need to reauth
    assert res == expected
    assert not mocked_run_login_flow.called
    assert not mocked_get_funcx_web_client.called

    res = mc.upstream_call(None)  # case: now must reauth
    assert res == expected
    assert mocked_run_login_flow.called
    assert mocked_get_funcx_web_client.called
```

Reply:
This looks good to me. I'll incorporate it and push some fresh commits.